REST API - Features Create & Update endpoints

bttf opened the pull request:
This PR adds two endpoints for **Feature Flags**
- `POST /features` (create)
- `POST /features/:id` (update)

Changes
- New schemas: 
  - `PostFeaturePayload.yaml`
  - `UpdateFeaturePayload.yaml`
  - Under `postFeature/` subdirectory (necessary due to required fields differing vs. read operation)
    - `FeatureEnvironment.yaml`
    - `FeatureExperimentRule.yaml`
    - `FeatureForceRule.yaml`
    - `FeatureRolloutRule.yaml`
    - `FeatureRule.yaml`
- Adds following methods for creating and updating environment settings for a feature
  - `fromApiEnvSettingsRulesToFeatureEnvSettingsRules`
  - `createInterfaceEnvSettingsFromApiEnvSettings`
  - `updateInterfaceEnvSettingsFromApiEnvSettings`

### Screenshots
Create
<img width="1582" alt="Screenshot 2023-07-17 at 9 17 32 PM" src="https://github.com/growthbook/growthbook/assets/2374625/8c22ec47-fd62-4a31-a5ba-407d984ee61e">

Update
<img width="1582" alt="Screenshot 2023-07-17 at 11 27 41 PM" src="https://github.com/growthbook/growthbook/assets/2374625/f30b72aa-6ced-4faa-af91-2d6ba0a62b3c">

msamper commented on packages/back-end/src/api/features/updateFeature.ts:
```diff
@@ -0,0 +1,125 @@
+        : req.body.defaultValue ?? null;
+
+    const updates: Partial<FeatureInterface> = {
+      ...(owner != null ? { owner } : {}),
+      ...(archived != null ? { archived } : {}),
+      ...(description != null ? { description } : {}),
+      ...(project != null ? { project } : {}),
+      ...(tags != null ? { tags } : {}),
+      ...(defaultValue != null ? { defaultValue } : {}),
+      ...(environmentSettings != null ? { environmentSettings } : {}),
```
could consider using lodash isNull here to make a little cleaner but not a big deal

bttf replied:
Technically would have to use `isNil`. `!= null` translates to 'is not null or undefined'. But not using a 3rd party library is best IMO

bttf replied:
(`!=` !== `!==`)

igorlovich commented on packages/back-end/generated/spec.yaml:
```diff
@@ -205,6 +420,62 @@ paths:
                 properties:
                   feature:
                     $ref: '#/components/schemas/Feature'
+    post:
```
Should this be a PUT ? or a PATCH if partial updates are allowed

bttf replied:
We are establishing a convention to use `POST` for partial updates for simplicity

bttf commented on packages/back-end/src/api/features/postFeature.ts:
```diff
@@ -0,0 +1,157 @@
+      environmentSettings,
+    };
+
+    req.checkPermissions(
```
We could but this is preferable - we're leveraging the `feature` object in `checkPermissions` call. If creating a POJO were an expensive operation, I could see the argument for preventing it but in this case it helps readability

bttf commented on packages/back-end/src/api/openapi/payload-schemas/PostFeaturePayload.yaml:
```diff
@@ -0,0 +1,46 @@
+      - type: string
+      - type: number
+      - type: boolean
+      - type: object
```
`type: string`

bttf replied:
consolidate to `type: string`

bttf commented on packages/back-end/src/services/features.ts:
```diff
@@ -839,3 +855,91 @@ function sha256(str: string, salt: string): string {
+            value: v.value,
+            name: v.name ?? "",
+            weight: v.weight,
+          })) ?? [],
```
These need to be validated

bttf commented on packages/back-end/src/services/features.ts:
```diff
@@ -839,3 +855,91 @@ function sha256(str: string, salt: string): string {
+        id: r.id ?? "",
+        type: r.type,
+        description: r.description ?? "",
+        value: r.value,
```
Here

bttf commented on packages/back-end/src/services/features.ts:
```diff
@@ -839,3 +855,91 @@ function sha256(str: string, salt: string): string {
+      coverage: r.coverage,
+      description: r.description ?? "",
+      hashAttribute: r.hashAttribute,
+      value: r.value,
```
Here

bttf commented on packages/back-end/src/api/features/postFeature.ts:
```diff
@@ -0,0 +1,157 @@
+        },
+      },
+      jsonSchema: {
+        schema: "",
```
- Accept jsonSchema input
- Set if org is enterprise
- Ignore if not
- Mention in docs

bttf commented on packages/back-end/src/api/features/updateFeature.ts:
```diff
@@ -0,0 +1,126 @@
+
+    // ensure default value matches value type
+    if (req.body.defaultValue != null) {
+      validateDefaultValueType(feature.valueType, req.body.defaultValue);
```
This will change

bttf commented on packages/back-end/src/api/openapi/payload-schemas/postFeature/FeatureRule.yaml:
```diff
@@ -0,0 +1,4 @@
+anyOf:
+  - $ref: "./FeatureForceRule.yaml"
+  - $ref: "./FeatureRolloutRule.yaml"
+  - $ref: "./FeatureExperimentRule.yaml"
```
FeatureExperimentRule will be nixed out; better to wait after #1139 merges

igorlovich commented:
> PR is blocked by #1139 & #1525

Both of these have been merge, should the blocked label be removed ?

igorlovich commented:
@bttf what is the priority of this item in your backlog ?

bttf commented on packages/back-end/src/api/features/postFeature.ts:
```diff
@@ -0,0 +1,151 @@
+    date: new Date(),
+    enabled: false,
+  };
+  const commercialFeatures = [...accountFeatures[getAccountPlan(org)]];
+  if (!commercialFeatures.includes("json-validation")) return defaultJsonSchema;
```
@jdorn I could use a confirmation on whether I'm checking for enterprise features correctly here

bttf commented:
@jdorn Following our conversation re: license key checks, the [most recent commit](https://github.com/growthbook/growthbook/pull/1478/commits/5f9f1a974fb0b8af9167cda4894f77d05ce141df) makes the following changes so that we always enforce license key checking for both internal API requests and REST API requests.

1. In `processJWT` (our auth middleware responsible for loading req.organization for the internal API), I've added a `licenseInit` call that supplies `req.organization.licenseKey`, which will be set only if the org has the license key stored on its mongoDB row. Otherwise, licenseInit should fallback to the env var, and simply bypass if none is found
2. With `licenseInit`, the subsequent `verifyLicenseMiddleware` should now always verify either env-var OR mongoDB sourced license keys
3. For the REST API, I've added a similar `licenseInit` call to the middleware responsible for loading the organization based on the API Key
4. I've also added subsequent `verifyLicenseMiddleware` to the REST API middleware chain
5. With `verifyLicenseMiddleware` in the REST API chain, we can now check for `res.locals.licenseError` in our REST API endpoints. I've added both checks to `postFeature` and `updateFeature` accordingly

bttf commented:
Tested creating an experiment ref rule
<img width="1624" alt="Screenshot 2023-10-06 at 11 31 50 AM" src="https://github.com/growthbook/growthbook/assets/2374625/1d648c6f-295b-4e67-ad26-fcf3cbd1e481">

jdorn commented on packages/back-end/src/api/features/postFeature.ts:
```diff
@@ -0,0 +1,155 @@
+  const commercialFeatures = [...accountFeatures[getAccountPlan(org)]];
+  if (!commercialFeatures.includes("json-validation")) return defaultJsonSchema;
+  try {
+    return jsonSchema ? JSON.parse(jsonSchema) : defaultJsonSchema;
```
Instead of the user passing in an object like this (requiring nested JSON stringify and date/enabled):
```js
JSON.stringify({
  "schema": JSON.stringify(schema),
  date: new Date(),
  enabled: true
})
```

I think we can simplify it if they only pass in `JSON.stringify(schema)` itself.  We can auto-populate `date` and `enabled` for them.  I can't think of a good reason why someone would want to pass in a JSON schema, but set `enabled` to false

bttf replied:
Good catch, updated

jdorn commented on packages/back-end/src/api/features/updateFeature.ts:
```diff
@@ -0,0 +1,140 @@
+      ...(jsonSchema != null ? { jsonSchema } : {}),
+    };
+
+    if (updates.environmentSettings) {
```
There are some other fields that require this permission since it can change the published SDK payloads:
- defaultValue
- project
- archived

bttf replied:
Updated

jdorn commented on packages/back-end/src/services/features.ts:
```diff
@@ -906,3 +923,90 @@ export function sha256(str: string, salt: string): string {
+  incomingEnvs: ApiFeatureEnvSettings
+): FeatureInterface["environmentSettings"] => {
+  const existing = feature.environmentSettings;
+  return Object.keys(incomingEnvs).reduce((acc, k) => {
```
What happens if the feature currently has environmentSettings for both `dev` and `production`, but the update call only specifies changes for `production`?  If I'm following the logic right, I think this will delete the existing `dev` settings, which we don't want to do.  Is that right?

bttf replied:
If a feature has envSettings for both `dev` and `production`, and the update only includes `production`, it will not override `dev`. We use `existing` to hold existing envSettings and use that as the initial value for `acc` in the reduce fn, so we only overwrite what is supplied by the update. Additionally, if the update only includes the `enabled` field for an envSetting (and doesn't specify the rules), we **do not** erase the rules but include the existing set of rules with the new `enabled` value ... Let me know if that makes sense

https://github.com/growthbook/growthbook/blob/66f92e3deee419865ed3dc4828675e1d7d0e8f50/packages/back-end/src/services/features.ts#L998-L1011

jdorn commented on packages/back-end/src/api/features/postFeature.ts:
```diff
@@ -42,19 +42,22 @@ export const parseJsonSchemaForEnterprise = (
     schema: "",
     date: new Date(),
-    enabled: false,
+    enabled: true,
```
Shouldn't this default to false and only get set to true if a schema is supplied and the org is enterprise?

bttf replied:
That was a silly mistake. Fixed

bttf commented:
@igorlovich This is now live.
